handlers_test.go

```diff
@@ -0,0 +1,128 @@
+package mockoidc
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"strconv"
+	"strings"
+	"testing"
+
+	"github.com/dgrijalva/jwt-go"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestAuthorizeHandler(t *testing.T) {
+
+	keypair, _ := RandomKeypair(2048)
+	m, _ := NewServer(keypair.PrivateKey)
+
+	data := url.Values{}
+	data.Set("scope", "testScope")
+	data.Set("response_type", "json?")
+	data.Set("redirect_uri", "example.com")
+	data.Set("state", "testState")
+	data.Set("client_id", m.ClientID)
+	assert.HTTPError(t, m.Authorize, "GET", authorizeEndpoint, nil)
+
+	// valid request
+	assert.HTTPStatusCode(t, m.Authorize, "GET", authorizeEndpoint, data, 302)
+
+	// Bad client ID
+	data.Set("client_id", "wrong_id")
+	assert.HTTPStatusCode(t, m.Authorize, "GET", authorizeEndpoint, data, 401)
+	assert.HTTPBodyContains(t, m.Authorize, "GET", authorizeEndpoint, data, invalidClient)
+
+	// Missing form value -- scope
+	for key, _ := range data {
+		newData := cloneValues(data)
+		newData.Del(key)
+		assert.HTTPStatusCode(t, m.Authorize, "GET", authorizeEndpoint, *newData, 400)
+		assert.HTTPBodyContains(t, m.Authorize, "GET", authorizeEndpoint, *newData, invalidRequest)
+	}
+}
+
+func TestTokenAuthorization(t *testing.T) {
+	keypair, _ := RandomKeypair(2048)
+	m, _ := NewServer(keypair.PrivateKey)
+	session, _ := m.SessionStore.NewSession("sessionScope", "sessionStrate", "sessionNonce", DefaultUser())
+
+	assert.HTTPError(t, m.Token, "POST", tokenEndpoint, nil)
+
+	data := url.Values{}
+	data.Set("client_id", m.ClientID)
+	data.Set("client_secret", m.ClientSecret)
+	data.Set("code", session.SessionID)
+	data.Set("grant_type", "authorization_code")
+
+	// all values must be provided
+	for key, _ := range data {
+		newData := cloneValues(data)
```

Comment:
I wonder if `url.Values.Encode` -> `url.ParseQuery` is an easier way to clone using purely the library?

Reply:
Alternatively you can make a helper function that intializes your `url.Value` with those defaults you want for testing and you just make fresh copies here.